packages/lib/gpcircuits/src/tuple.ts

```diff
@@ -0,0 +1,171 @@
+ * Throws a `RangeError` if the tuple arity parameter is not proper.
+ */
+export function requiredNumTuples(
+  paramTupleArity: number,
+  tupleArity: number
+): number {
+  if (paramTupleArity < 2) {
+    throw new RangeError("The tuple arity parameter must be at least 2.");
+  } else {
+    return tupleArity <= paramTupleArity
+      ? 1
+      : Math.ceil((tupleArity - 1) / (paramTupleArity - 1));
+  }
+}
+
+/**
+ * Determines the maximum tuple arity representable by the
+ * given parameters.
+ * Throws a `RangeError` is the parameters are not proper.
+ */
+export function maxTupleArity(
+  paramMaxTuples: number,
+  paramTupleArity: number
+): number {
+  if (paramTupleArity < 2) {
+    throw new RangeError("The tuple arity parameter must be at least 2.");
+  } else if (paramMaxTuples < 0) {
+    throw new RangeError(
+      "The maximum tuple number parameter must be nonnegative."
+    );
+  } else {
+    return paramMaxTuples === 0
+      ? 0
+      : paramTupleArity + (paramTupleArity - 1) * (paramMaxTuples - 1);
+  }
+}
+
+/**
+ * This procedure takes an N-tuple of `indices` (referring
+ * to elements of some array of values) and returns its
+ * representation as an appropriately linked sequence of
+ * tuples of arity `paramMaxArity`, where these tuples
+ * are indexed starting at `firstTupleIndex` and the
+ * first index is used as padding. More concretely,
+ * Given indices [i_1, ..., i_N], if N <= `paramTupleArity`,
+ * computeTupleIndices(paramTupleArity, firstTupleIndex, indices)
+ *  === padArray([i_1, ..., i_N], paramTupleArity, i_1), else
+ * computeTupleIndices(paramTupleArity, firstTupleIndex, indices)
+ * === [indices.slice(0, paramTupleArity)]
+ *		.concat(tupleIndices(
+ *				     paramTupleArity,
+ *				    firstTupleIndex + 1,
+ *				    [firstTupleIndex].concat(
+				 indices.slice(paramTupleArity)))).
+ * Examples:
+ * computeTupleIndices(2, 5, [1, 3, 4]) === [[1, 3], [5, 4]]
+ * computeTupleIndices(3, 5, [0, 1, 4, 2]) === [[0, 1, 4], [5, 2, 0]]
+ * computeTupleIndices(4, 6, [3, 4, 2, 1, 5]) === [[3, 4, 2, 1], [6, 5, 3, 3]]
+ *
+ * @param paramTupleArity the arity of the output tuple
+ * @param firstTupleIndex the index of the first output tuple in the combined array of
+ * entry and tuple value hashes
+ * @param indices the entry value indices forming the input tuple
+ * @returns list of index tuples of arity `paramTupleArity` representing the input tuple
+ * @throws RangeError if the parameters are not in the proper ranges.
+ */
+export function computeTupleIndices(
+  paramTupleArity: number,
+  firstTupleIndex: number,
+  indices: number[]
+): number[][] {
+  // Verify that the parameters make sense
+  if (paramTupleArity < 2 || paramTupleArity > 4) {
+    throw new RangeError("The tuple arity must lie between 2 and 4.");
+  }
+  if (indices.some((i) => i >= firstTupleIndex)) {
+    throw new RangeError(
+      "The first tuple index must be strictly greater than the given indices."
```

Comment:
We should think at some point about how to handle tuples manually formed by users in the config.  That'll be relevant when there are more places for a tuple to go than just into lists.  I think it's fine not to worry about it yet, though.